src/ekam/rules/intercept.c

```diff
@@ -98,7 +100,30 @@ static pthread_mutex_lock_func* dynamic_pthread_mutex_lock = NULL;
 static pthread_mutex_unlock_func* dynamic_pthread_mutex_unlock = NULL;
 
 int fake_pthread_once(pthread_once_t* once_control, void (*init_func)(void)) {
+  static pthread_mutex_t fake_once_mutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
Did you actually observe a problem with pthread_once() or did you find this issue by reading the code?

The idea behind `init_pthreads()` is that if libpthread is linked in, then it should find the real implementations of the functions, but if not, then we assume the application is single-threaded and therefore thread-unsafe initialization is fine.

But then there's the comment:

```
/* TODO:  For some reason the pthread_once returned by dlsym() doesn't do anything? */
```

I vaguely remember, I observed that `dlsym("pthread_once")` did find a function, but calling that function had no effect -- in particular, it did _not_ run the init callback that it was given. It looks like I worked around that problem by always using the fake pthread_once() that is not actually safe, adding `|| 1` to the conditional.

That was like a million years ago, though. Did you try removing the `|| 1` to see if that problem is still present?

I would _assume_ that in the case where `dlsym()` returned a `pthread_once()` that did nothing, then it probably also returns `pthread_mutex_lock()` and `pthread_mutex_unlock()` functions that do nothing. But in these cases, the "fake" versions also do nothing (that's the correct thing to do in the single-threaded case), so it didn't create any actual problem.

But that means that, in this case, your "fixed" pthread_once actually doesn't work either, because it relies on `dynamic_pthread_mutex_lock`, which... does nothing.

Reply:
I did observe a problem in that the "once" functor executed many times. I can't say if it's responsible for any race conditions for sure since I'd have to TSAN the interceptor & that's a mild PITA, but I struggle to see any situation where that code isn't slightly risky/broken.

I haven't tried removing the `|| 1` as I assumed there was a good reason for it (maybe `pthread_once` can't actually be intercepted due to how glibc is built?).

If `pthread_mutex_lock` doesn't work or isn't available, `fake_pthread_once` should still ensure the init_func is invoked just once. If it does work, then it's actually thread safe. There may be a corner condition where init_pthreads doesn't initialize because the program wasn't linked against pthreads - I may need to double-check that.

Reply:
https://github.com/namhyung/uftrace/issues/714 would seem to indicate that `pthread_once` itself does a dlsym which may explain why intercepting it is problematic.

Reply:
We're not intercepting pthread_once, we're just trying to call it.

Can you try removing the `|| 1` and see if it's still broken? If it works then we can avoid changing anything.

If it doesn't work then the pthread_mutex in your fake probably doesn't work either, but I do see your point that the old fake implementation didn't actually ensure that the init was only called once... and it just happened to be that the one init function it was used on was idempotent. Oops. I do agree the fake should at least be fixed to avoid that.

Another thing we could try here is doing the initialization as a global constructor (use `__attribute__((constructor))`) which would mean we can avoid pthreads altogether... however, that only works if the dynamic linker actually initializes this module before any other module has a chance to make syscalls that need to be intercepted.

Reply:
That all sounds like a lot of work to validate & I've already killed a few hours trying to do the execvpe approach which turns out to be a dead end. I tested the `|| 1` on my Arch machine & it doesn't appear to have the issue the comments do but I have no idea if this might be an issue on Buster. I've already spent more time on this patch than I intended so my preference would be to just leave TODOs for another time unless you're 100% certain there's no possible issues removing the `||1`